### PR TITLE
fix(subs mgmt api): attempt to return mixed clients ds/non-ds in the same page

### PR DESCRIPTION
**Pull request by thalesmg**
Fixes https://emqx.atlassian.net/browse/EMQX-12294

Release version: v/e5.7

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

File: apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl [hunk 1]
```diff
@@ -396,10 +401,21 @@ merge_queries(QString0, Q1, Q2) ->
             Q2Page = ceil(C1 / Limit),
             case Page =< Q2Page of
                 true ->
-                    #{data := Data, meta := #{hasnext := HN}} = Q1(QString0),
+                    #{data := Data1, meta := #{hasnext := HN1}} = Q1(QString0),
+                    NumRows1 = length(Data1),
+                    case NumRows1 >= Limit of
+                        true ->
+                            Data = Data1,
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl:408]: this kind of variable bind is very confusing, it looks like assertion

suggestion: `{Data, HasNext} = case ... of`
or even better, write a func to handle the case part.

Comment (thalesmg) [reply apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl]: :heavy_check_mark:

File: apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl [hunk 2]
```diff
@@ -310,7 +310,12 @@ consume_n_matching(Map, Pred, N, S) ->
 consume_n_matching(_Map, _Pred, _N, [], Acc) ->
     {lists:reverse(Acc), []};
 consume_n_matching(_Map, _Pred, 0, S, Acc) ->
-    {lists:reverse(Acc), S};
+    case emqx_utils_stream:next(S) of
+        [] ->
+            {lists:reverse(Acc), []};
```
Comment (savonarola) [inline apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl:315]:
```suggestion
            {lists:reverse(Acc), emqx_utils_stream:empty()};
```
?

File: apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl [hunk 3]
```diff
@@ -421,6 +431,31 @@ merge_queries(QString0, Q1, Q2) ->
...
+    } = Params,
+    NumRows1 = length(Data1),
+    {Data, HN} =
+        case NumRows1 >= Limit of
+            true ->
+                {Data1, HN1 orelse C2 > 0};
+            false ->
+                #{data := Data2, meta := #{hasnext := HN2}} =
+                    Q2(QString0#{<<"limit">> := Limit - NumRows1}),
+                {Data1 ++ Data2, HN1 or HN2}
```
Comment (savonarola) [inline apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl:452]: Curious: if we assume that HN1 may be true, why do we use the second query?

Comment (thalesmg) [reply apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl:452]: In this arm of the `case`, the number of returned rows from `Q1` is fewer than `Limit`, so I think we might assume `HN1 = false` in most cases.  But I think you have a good point: this situation is not necessarily wrong, I'll check for this condition.  Thanks!  :beers:

File: apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl (hunk 2 above)
Comment (thalesmg) [reply apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl:315]: Actually, that's part of the fix: there is this check to determine if `hasnext = true`: https://github.com/thalesmg/emqx/blob/0dee2d67b1645b2ae2bcb46d2817c56b03b89f6f/apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl#L258

File: apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl (hunk 3 above)
Comment (thalesmg) [reply apps/emqx_management/src/emqx_mgmt_api_subscriptions.erl]: :heavy_check_mark:

**Review by zmstone — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution]()

**Pull request merged**